npm/private/npm_translate_lock_state.bzl

```diff
@@ -0,0 +1,398 @@
+    if utils.exists(rctx, action_cache_path):
+        input_hashes = parse_npmrc(rctx.read(action_cache_path))
+        if utils.dicts_match(input_hashes, priv["input_hashes"]):
+            # Calculated input hashes match saved input hashes; nothing to update
+            return False
+    return True
+
+########################################################################################################################
+def _write_action_cache(priv, rctx, label_store):
+    contents = [
+        "# Input hashes for repository rule: npm_translate_lock(name = \"{}\", pnpm_lock = \"{}\")".format(rctx.name, label_store.label("pnpm_lock")),
+    ]
+    for key, value in priv["input_hashes"].items():
+        contents.append("{}={}".format(key, value))
+    rctx.file(
+        label_store.repository_path("action_cache"),
+        "\n".join(contents) + "\n",
+    )
+    utils.reverse_force_copy(
+        rctx,
+        label_store.label("action_cache"),
+        label_store.path("action_cache"),
+    )
+
+########################################################################################################################
+def _copy_input_file(priv, rctx, label_store, key):
+    if not label_store.has(key):
+        fail("key not found '{}'".format(key))
+    contents = rctx.read(label_store.path(key))
+    if _update_pnpm_lock(priv):
+        _set_input_hash(
+            priv,
+            label_store.relative_path(key),
+            utils.hash(contents),
+        )
+    rctx.file(
+        label_store.repository_path(key),
+        content = contents,
+        executable = False,
+    )
+
+########################################################################################################################
+def _load_npmrc(priv, rctx, label_store):
+    npmrc_path = label_store.path("npmrc")
+
+    contents = parse_npmrc(rctx.read(npmrc_path))
+    if "registry" in contents:
+        priv["default_registry"] = utils.to_registry_url(contents["registry"])
+
+    (registries, tokens, basic_auth) = helpers.get_npm_auth(contents, npmrc_path, rctx.os.environ)
+    priv["npm_registries"] = registries
+    priv["npm_tokens"] = tokens
+    priv["npm_basic_auth"] = basic_auth
+
+########################################################################################################################
+def _load_lockfile(priv, rctx, label_store):
+    importers, packages = utils.parse_pnpm_lock(rctx.read(label_store.path("pnpm_lock")))
+    priv["importers"] = importers
+    priv["packages"] = packages
+
+########################################################################################################################
+def _has_workspaces(priv):
+    importer_paths = priv["importers"].keys()
+    return importer_paths and (len(importer_paths) > 1 or importer_paths[0] != ".")
+
+########################################################################################################################
+def _update_pnpm_lock(priv):
```

Comment:
"update" seems like an action would be performed. I wonder if these getter functions ought to be prefixed with get_

Reply:
I like the golang guidance of not prefixing getters with `Get`, just setters with `Set`

Reply:
better to rename the value if it is confusing to something like `should_update_pnpm_lock`

Reply:
https://go.dev/doc/effective_go#Getters for refere